### PR TITLE
Support mounting react elements in editor

### PR DESCRIPTION
**Pull request by Adjective-Object**
Migrating here from microsoft/roosterjs/pull/360
---

### Changes
- Adds a plugin to support mounting custom react elements in-editor
- Adds a navbar to the sample sites for easier navigating between them when running multiple examples.
- Adds `karma` and uses it to run integration tests on the react plugin

---

### Goals:
- components should be copy/pasteable
- A should be able to be drag/dropped around the Dom without losing state
- components should be able to be deleted & reinserted via undo/redo, preserving a reasonable amount of internal state
- consumers using this plugin should have to be minimally aware of the fact that they're not mounting in the same DOM that is being rendered to the user.

---

### Approach

In order for darkmode to interoperate with custom react components in the editor, their transforms need to go through some darkmode-aware content insertion pipeline.

This proposal works by tracking component mountpoints across the editor surface, and mirroring components mounted off the actual document root into the discovered mountpoints. This option was chosen over direct mounting or a custom react reconciler for the following reasons:

- direct mounting: rerenders of the component can't be tracked and be made darkmode aware, nor can their updates be tracked by our undo buffer (if that's something we care to do).
- react reconciler: this seems like the best way in theory, but the additional complexity & code doesn't seem worth the tradeoff, and the react-reconciler api isn't stable.

---
### Test plan:

`yarn karma start`
`yarn start`, and play with the experimental custom component (copy/paste the component into the editor from its section in the sidebar).

---

### Future Work
- I want to look at using a custom react-reconciler to:
  - avoid the consumer needing to call `updateDomInEditor` when their component updates.
  - make it possible to diff/patch updates to the dom instead of forcing a full update of the component's inner boundary every time
- I want to look at avoiding the wrapper span, and instead mounting the root rendered component w/ additional data fields directly in the editor.

### CODE REVIEW

File: packages/roosterjs-plugin-react/lib/util/removeReactTagsOnExtractContent.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+} from './constants';
+
+const simpleReactAttributes = new RegExp(
+    `(${REACT_COMPONENT_DATA_KEY}|${REACT_COMPONENT_INSTANCE_ID}|${REACT_COMPONENT_SHARABLE_STATE})=("[^"]*"|'[^']*'|[^\t\n\r\f ><"'\`]*)`,
+    'gim'
+);
+
+export default function removeReactTagsOnExtractContent(html: string): string {
+    if (html) {
+        const withoutSimpleTags = html.replace(simpleReactAttributes, '');
```
Comment (JiuqingSong) [inline packages/roosterjs-plugin-react/lib/util/removeReactTagsOnExtractContent.ts]: We should only filter these strings within an HTML tag, not the content

File: sample/script/FocusOutShellSample.tsx [hunk 2]
```diff
@@ -223,6 +224,7 @@ function createEditor(name: string, loadEmojiStrings: boolean = false): JSX.Elem
 
 const view = (
     <div className="root-container">
+        <NavBar />
```
Comment (JiuqingSong) [inline sample/script/FocusOutShellSample.tsx]: @bochong please make sure you review this.

File: packages/roosterjs-plugin-react/lib/ReactPlugin.tsx [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+
+    /**
+     * Check if the plugin should handle the given event exclusively.
+     * Handle an event exclusively means other plugin will not receive this event in
+     * onPluginEvent method.
+     * If two plugins will return true in willHandleEventExclusively() for the same event,
+     * the final result depends on the order of the plugins are added into editor
+     * @param _event The event to check
+     */
+    public willHandleEventExclusively(_event: PluginEvent) {
```
Comment (JiuqingSong) [inline packages/roosterjs-plugin-react/lib/ReactPlugin.tsx]: You can remove this function

File: packages/roosterjs-plugin-react/lib/ReactPlugin.tsx [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+    private handleChangeInEditor(event: PluginEvent) {
+        if (!this.editor) {
+            return;
+        }
+
+        // Undoes and other major changes to document content fire this type of event.
+        //
+        // Scan for new elements.
+        const foundComponentRoots = this.editor.queryElements(`[${REACT_COMPONENT_DATA_KEY}="${this.componentId}"]`);
```
Comment (JiuqingSong) [inline packages/roosterjs-plugin-react/lib/ReactPlugin.tsx]: Does this mean multiple react components share the same component id?

**Comment by bochong** [comment]
@JiuqingSong This may be an issue from the previous PR updating react, but when i run "npm run build", it seems to be failing with some errors. They seem like minor errors, so can we fix those in this PR as well?

File: packages/roosterjs-plugin-react/lib/ReactPlugin.tsx (hunk 4 above)
Comment (Adjective-Object) [reply packages/roosterjs-plugin-react/lib/ReactPlugin.tsx]: the `componentId` here, is an identifier of the react component type (e.g. is shared by all components with the same component constructor, or component function).

This is so you can host multiple types of components in the rooster js editor.

I'll rename `componentTypeId` in the class and document it

File: packages/roosterjs-plugin-react/lib/util/removeReactTagsOnExtractContent.ts (hunk 1 above)
Comment (Adjective-Object) [reply packages/roosterjs-plugin-react/lib/util/removeReactTagsOnExtractContent.ts]: Sounds good.

Should I hold with this string scanning approach, or do you think I ought to parse the string to a DOM? I was a little worried about the perf impact of parsing & re-serializing the HTML string

**Comment by JiuqingSong** [comment]
> @JiuqingSong This may be an issue from the previous PR updating react, but when i run "npm run build", it seems to be failing with some errors. They seem like minor errors, so can we fix those in this PR as well?

https://github.com/microsoft/roosterjs-react/pull/102

File: packages/roosterjs-plugin-react/lib/util/removeReactTagsOnExtractContent.ts (hunk 1 above)
Comment (JiuqingSong) [reply packages/roosterjs-plugin-react/lib/util/removeReactTagsOnExtractContent.ts]: I think we can combine the regex search and simple string search to optimize for perf.

Here's what I did for image proxy, it firstly match with a regex, then check if the match result is inside a tag or not:

            bool replaced = false;
            char[] charToSearch = new char[] { '<', '>' };

            content = WebImageController.scriptRegex.Replace(
                content,
                match =>
                {
                    replaced = true;
                    return string.Empty;
                });

            content = WebImageController.onEventRegex.Replace(
                content,
                match =>  
                {
                    int c = content.LastIndexOfAny(charToSearch, match.Index - 1);
                
                    // Check if the match is inside a tag
                    if (content[c] == '<')
                    {
                        replaced = true;
                        return string.Empty;
                    }
                    else
                    {
                        return match.Value;
                    }
                });

            newContent = content;
            return replaced;

**Pull request closed without merging**